File: nvpl_blas/c/dgbmv.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_dgbmv
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t M = 3;
    nvpl_int_t N = 4;
    nvpl_int_t KL = 1;
    nvpl_int_t KU = 2;
    double alpha = 1.0f;
    double beta = 1.0f;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_TRANSPOSE transA = CblasNoTrans;
    double * A = NULL;
    double * X;
    double * Y;
    nvpl_int_t incX = 1;
    nvpl_int_t incY = 1;

    nvpl_int_t nx = (transA == CblasNoTrans) ? N : M;
    nvpl_int_t ny = (transA == CblasNoTrans) ? M : N;
    
    nvpl_int_t lda = KL + KU + 1;
    nvpl_int_t size_k = (order == CblasColMajor) ? N : M;
    nvpl_int_t len_x = 1 + (nx - 1) * labs(incX);
    nvpl_int_t len_y = 1 + (ny - 1) * labs(incY);

    printf("\nExample: cblas_dgbmv for banded matrix-vector multiplication\n\n");
    printf("#### args: m=%" PRId64 ", n=%" PRId64 ", kl=%" PRId64 ", ku=%" PRId64 ", lda=%" PRId64 ", "
            "incx=%" PRId64 ", incy=%" PRId64 ", transA=%c, order=%c\n", (int64_t)M, (int64_t)N, (int64_t)KL,
           (int64_t)KU, (int64_t)lda, (int64_t)incX, (int64_t)incY, transpose_to_char(transA),
           order_to_char(order));
    printf("           alpha=%g, beta=%g\n", alpha, beta);

    // allocate memory
    A = (double *)malloc(lda * size_k * sizeof(double));
    X = (double *)malloc(len_x * sizeof(double));
    Y = (double *)malloc(len_y * sizeof(double));

    // fill data
    fill_dbandmatrix(A, KL, KU, M, N, lda, order);
    fill_dvector(X, nx, incX);
    fill_dvector(Y, ny, incY);

    // print input data
    print_dbandmatrix(A, KL, KU, M, N, lda, "A", order);
    print_dvector(X, nx, incX, "X");
    print_dvector(Y, ny, incY, "Y");

    // call cblas_dgbmv
    cblas_dgbmv(order, transA, M, N, KL, KU, alpha, A, lda, X, incX, beta, Y, incY);

    // print result
    printf("\nAfter dgbmv operation:\n");
    print_dvector(Y, ny, incY, "Y");

    // release memory
    free(A);
    free(X);
    free(Y);
    return 0;
}

File: nvpl_blas/f77/dgbmv.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     DGBMV 
!
!******************************************************************************
      program   DGBMV_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      character*1      transa
      parameter        (transa='N')
      integer          m, n, lda
      integer          incx, incy, i, nx, ny, xymax
      parameter        (m=3, n=4, incx=1, incy=1, xymax=4)
      integer          kl, ku
      parameter        (kl=1, ku=2, lda=kl+ku+1)
      integer          asize
      parameter        (asize=m*n)
      double precision alpha, beta
      parameter        (alpha=1.0, beta=1.0)
      double precision a(lda,n)
      integer          xsize, ysize
      parameter        (xsize=1+(xymax-1)*abs(incx),
     &                  ysize=1+(xymax-1)*abs(incy))
      double precision x(xsize), y(ysize)

*       External Subroutines
      external        print_dbandmatrix, fill_dbandmatrix,
     &                print_dvector, fill_dvector
*
*      Executable Statements
*
      print*
      print 99, 'Example: DGBMV for banded matrix-vector ',
     &          'multiplication'
*
*       Print input data
      print*
      print 100, m, n, kl, ku, lda, incx, incy, transa
      print 101, alpha, beta
      if (transa .eq. 'N') then
          nx=n
          ny=m
      else
          nx=m
          ny=n
      endif
      call fill_dbandmatrix(a,kl,ku,m,n,lda)
      call fill_dvector(x,nx,incx)
      call fill_dvector(y,ny,incy)
      call print_dbandmatrix(a,kl,ku,m,n,lda,'A')
      call print_dvector(x,nx,incx,'X')
      call print_dvector(y,ny,incy,'Y')
*
*      Call DGBMV subroutine
      call DGBMV(transa,m,n,kl,ku,alpha,a,lda,x,incx,beta,y,incy)
*
      print*
      print 102, 'After dgbmv operation:'
      call print_dvector(y,ny,incy,'Y')

      stop
 99   format(A,A)
 100  format('#### args: m=',i1,', n=',i1,', kl=',i1,', ku=',i1,
     &       ', lda=',i1,', incx=',i1,', incy=',i1,', transA=',a1)
 101  format(11x,'alpha=',f4.1,'  beta=',f4.1)
 102  format(A)
      end
